src/php/lib/Grpc/BaseStub.php

```diff
@@ -196,17 +211,184 @@ private function _validate_and_normalize_metadata($metadata)
+                    $this->update_metadata,
+                $metadata,
+                $jwt_aud_uri
+                );
+            }
+            $metadata = $this->_validate_and_normalize_metadata(
+                $metadata
+            );
+            $call->start($argument, $metadata, $options);
+            return $call;
+        };
+    }
+
+    private function GrpcStreamUnary($channel)
+    {
+        return function ($method, $deserialize, array $metadata = [], array $options = []) use ($channel) {
+            $call = new ClientStreamingCall(
+                $channel,
+                $method,
+                $deserialize,
+                $options
+            );
+            $jwt_aud_uri = $this->_get_jwt_aud_uri($method);
+            if (is_callable($this->update_metadata)) {
+                $metadata = call_user_func(
+                    $this->update_metadata,
+                $metadata,
+                $jwt_aud_uri
+                );
+            }
+            $metadata = $this->_validate_and_normalize_metadata(
+                $metadata
+            );
+            $call->start($metadata);
+            return $call;
+        };
+    }
+
+    private function GrpcUnaryStream($channel)
+    {
+        return function ($method, $argument, $deserialize, array $metadata = [], array $options = []) use ($channel) {
+            $call = new ServerStreamingCall(
+                $channel,
+                $method,
+                $deserialize,
+                $options
+            );
+            $jwt_aud_uri = $this->_get_jwt_aud_uri($method);
+            if (is_callable($this->update_metadata)) {
+                $metadata = call_user_func(
+                    $this->update_metadata,
+                $metadata,
+                $jwt_aud_uri
+                );
+            }
+            $metadata = $this->_validate_and_normalize_metadata(
+                $metadata
+            );
+            $call->start($argument, $metadata, $options);
+            return $call;
+        };
+    }
+
+    private function GrpcStreamStream($channel)
+    {
+        return function ($method, $deserialize, array $metadata = [], array $options = []) use ($channel) {
+            $call = new BidiStreamingCall(
+                $channel,
+                $method,
+                $deserialize,
+                $options
+            );
+            $jwt_aud_uri = $this->_get_jwt_aud_uri($method);
+            if (is_callable($this->update_metadata)) {
+                $metadata = call_user_func(
+                    $this->update_metadata,
+                $metadata,
+                $jwt_aud_uri
+                );
+            }
+            $metadata = $this->_validate_and_normalize_metadata(
+                $metadata
+            );
+            $call->start($metadata);
+
+            return $call;
+        };
+    }
+
+    private function UnaryUnaryCallFactory($channel)
+    {
+        if (is_a($channel, 'Grpc\Channel')) {
```

Comment:
I would write the check for this function and the other three functions the opposite way:

```
if (is_a($channel, 'Grpc\InterceptorChannel')) {
   return function...
   ...
}
return this->GrpcUnaryUnary($channel); // you could inline this from the GrpcUnaryUnary function,
                                       // as I think it's the only call site.
```

The advantage is we are basically special-casing `InterceptorChannel` as this function knows how to treat them differently. For all other channel types (there is only one today, but in principle, should you want to mock the Channel object), we hand it over to what we have been doing today. That way, your code automatically support it, and also, when you are calling `getInterceptor()` you know you won't crash because the object definitely supports that method.